compiler/optimizer/LocalLiveRangeReducer.cpp

```diff
@@ -491,7 +491,7 @@ bool TR_LocalLiveRangeReduction::isAnySymInDefinedOrUsedBy(TR_TreeRefInfo *curre
       return true;
       }
 
-   if (comp()->target().cpu.isPower() && opCode.getOpCodeValue() == TR::allocationFence)
+   if (cg()->getEnforceStoreOrder() && opCode.getOpCodeValue() == TR::allocationFence)
```

Comment:
This change is obviously better since it does a more general check that includes AArch64.

A more general question I would have at this point is around why we need a check on enforcing store order here at all. i.e. is the presence of the allocation fence itself not a sign of store order needing to be enforced ?

Reply:
Yes

Reply:
There are some places in OpenJ9 that creates allocationFence nodes as shown below, and some of which checks `getEnforceStoreOrder()` before calling `createAllocationFence()`.

https://github.com/eclipse/openj9/blob/ee1f3d1ae159643f5cf8d3b3554224ee6282f082/runtime/compiler/ilgen/Walker.cpp#L7292
https://github.com/eclipse/openj9/blob/ee1f3d1ae159643f5cf8d3b3554224ee6282f082/runtime/compiler/optimizer/EscapeAnalysis.cpp#L6047
https://github.com/eclipse/openj9/blob/ee1f3d1ae159643f5cf8d3b3554224ee6282f082/runtime/compiler/optimizer/EscapeAnalysis.cpp#L8727
https://github.com/eclipse/openj9/blob/ee1f3d1ae159643f5cf8d3b3554224ee6282f082/runtime/compiler/optimizer/StringPeepholes.cpp#L144

I don't know what was the intention of the original code checking with `target().cpu.isPower()`.

Reply:
It would seem the better course of action is to not to insert the fences in the first place if the architecture doesn't require it.  Z and x86 both do nothing with allocation fences, P and AArch64 do, and AArch32 does nothing for allocation fences but perhaps it should...

This is obvious scope creep for this particular PR, but we may as well do the better thing now as it will ultimately solve this problem on AArch64 and other architectures coming in the future (e.g., RISC-V).

I also think that the capability query shouldn't be asked of the code generator but of the target CPU itself as that is really what's driving the need for the fences (i.e., a new query like `TR::Compiler->target.cpu.getEnforceStoreOrder()`)

Reply:
That also means that the `isPower()` guard in local live range reduction should simply be removed and the analysis applied any time it simply sees an allocation fence node.

Reply:
I removed the guards with `isPower()` and `getEnforceStoreOrder()`.

Reply:
I also opened https://github.com/eclipse/openj9/issues/10187.

Reply:
Thanks, no further questions from me. I have approved the change.